src/Asiago/Controllers/TwitchController.cs

```diff
@@ -0,0 +1,176 @@
+        private async Task<IActionResult> HandleEvent<T, I>(EventNotificationPayload<T> payload) where I : IInvocable, IInvocableWithPayload<EventNotificationPayload<T>>
+        {
+            VerifyMessageResult verifyResult = await VerifyMessageAsync();
+            if (!verifyResult.Verified)
+            {
+                return verifyResult.ErrorActionResult ?? BadRequest();
+            }
+
+            if (!verifyResult.Duplicate)
+            {
+                _queue.QueueInvocableWithPayload<I, EventNotificationPayload<T>>(payload);
+            }
+
+            return Ok();
+        }
+
+        /// <summary>
+        /// Record returned by <see cref="VerifyMessageAsync"/>.
+        /// </summary>
+        /// <param name="Verified">Whether the message is verified.</param>
+        /// <param name="Duplicate">Whether the message is a duplicate. Will always be false if <paramref name="Verified"/> is false since it won't be checked.</param>
+        /// <param name="ErrorActionResult">If <paramref name="Verified"/> is false, this should be returned by the endpoint that called <see cref="VerifyMessageAsync"/>. Otherwise it will be null.</param>
+        private record VerifyMessageResult(bool Verified, bool Duplicate, IActionResult? ErrorActionResult);
+
+        private async Task<VerifyMessageResult> VerifyMessageAsync()
+        {
+            IHeaderDictionary headers = HttpContext.Request.Headers;
+            if (!headers.TryGetValue(WebhookRequestHeaders.MessageId, out StringValues messageIdHeader)
+                || (string?)messageIdHeader is not string messageId
+                || !headers.TryGetValue(WebhookRequestHeaders.MessageTimestamp, out StringValues messageTimestampHeader)
+                || (string?)messageTimestampHeader is not string messageTimestamp
+                || !headers.TryGetValue(WebhookRequestHeaders.MessageSignature, out StringValues messageSignatureHeader)
+                || (string?)messageSignatureHeader is not string messageSignature)
+            {
+                return new(false, false, BadRequest("Missing required request headers."));
+            }
+
+            // Validate signature
+            HttpContext.Request.Body.Seek(0, SeekOrigin.Begin);
+            using (StreamReader reader = new(HttpContext.Request.Body))
+            {
+                string requestBody = await reader.ReadToEndAsync();
+                bool messageVerified = Utilities.VerifyMessageSignature(
+                    requestBody,
+                    messageId,
+                    messageTimestamp,
+                    messageSignature,
+                    _twitchOptions.WebhookSecret
+                    );
+                if (!messageVerified)
+                {
+                    _logger.LogError(
+                        "Signature validation failed for Twitch webhook request from [{ip}]",
+                        HttpContext.Connection.RemoteIpAddress

```

Comment:
When using a reverse proxy, this will not be the original ip address... look into if there's some way around this.

Reply:
https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.forwardedheadersextensions.useforwardedheaders?view=aspnetcore-8.0

something like this
```
// Configure ASP.NET Core to trust X-Forwarded-For header
app.UseForwardedHeaders(new ForwardedHeadersOptions
{
    ForwardedHeaders = Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedFor
});
```